core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -467,7 +467,9 @@ private[spark] class Master(
    * two executors on the same worker).
    */
   def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    worker.memoryFree >= app.desc.memoryPerSlave &&
+    !worker.hasExecutor(app) &&
+    !app.removedExecutors.exists(_.worker == worker)
```

Comment:
Yes, for long running applications, this will be a problem. I will re-implement it with TTL. Thanks @mateiz @markhamstra .